docs/proposals/horizontal-pod-autoscaler.md

```diff
@@ -0,0 +1,275 @@
+
+* Autoscaler avoids thrashing, i.e.: prevents rapid execution of conflicting decision if the load is not stable.
+
+
+As the CPU consumption of a pod immediately after start may be highly variable due to initialization/startup,
+autoscaler will skip metrics from the first minute of pod lifecycle.
+
+## Relative vs. absolute metrics
+
+The question arises whether the values of the target metrics should be absolute (e.g.: 0.6 core, 100MB of RAM)
+or relative (e.g.: 110% of resource request, 90% of resource limit).
+The argument for the relative metrics is that when user changes resources for a pod,
+she will not have to change the definition of the autoscaler object, as the relative metric will still be valid.
+However, we want to be able to base autoscaling on custom metrics in the future.
+Such metrics will rather be absolute (e.g.: the number of queries-per-second).
+Therefore, we decided to give absolute values for the target metrics in the initial version.
+
+Please note that when custom metrics are supported, it will be possible to create additional metrics
+in heapster that will divide CPU/memory consumption by resource request/limit.
+From autoscaler point of view the metrics will be absolute,
+although such metrics will be bring the benefits of relative metrics to the user.
+
+
+## Support in kubectl
+
+To make manipulation on HorizontalPodAutoscaler object simpler, we will add support for
+creating/updating/deletion/listing of HorizontalPodAutoscaler to kubectl.
+In addition, we will add kubectl support for the following use-cases:
+* When running an image with ```kubectl run```, there should be an additional option to create
+  an autoscaler for it.
+* When creating a replication controller or deployment with ```kubectl create [-f]```, there should be
+  a possibility to specify an additional autoscaler object.
+  (This should work out-of-the-box when creation of autoscaler is supported by kubectl as we may include
+  multiple objects in the same config file).
+* We will and a new command ```kubectl autoscale``` that will allow for easy creation of an autoscaler object
+  for already existing replication controller/deployment.
+
+## Future Features
+
+We list here some features that will not be supported in the initial version of autoscaler.
+However, we want to keep them in mind, as they will most probably be needed in future.
+Our design is in general compatible with them.
+* Autoscale pods based on metrics different than CPU & memory (e.g.: network traffic, qps).
+  This includes scaling based on a custom metric.
+* Autoscale pods based on multiple metrics.
+  If the target numbers of pods for different metrics are different, choose the largest target number of pods.
+* Scale the number of pods starting from 0: all pods can be turned-off,
+  and then turned-on when there is a demand for them.
+  When a request to service with no pods arrives, kube-proxy will generate an event for autoscaler
+  to create a new pod.
```

Comment:
Yeah, something akin to ```wake on lan``` (un-idling) would be of very useful for hosted/densely packed environments.